Applications/Spire/Source/TimeAndSales/TimeAndSalesTableView.cpp

```diff
@@ -0,0 +1,164 @@
+      int column) {
+    auto column_id = static_cast<TimeAndSalesTableModel::Column>(column);
+    auto cell = [&] () -> QWidget* {
+      if(column_id == TimeAndSalesTableModel::Column::TIME) {
+        return make_time_cell(table->get<ptime>(row, column));
+      } else if(column_id == TimeAndSalesTableModel::Column::PRICE) {
+        auto money_cell = make_label(to_text(table->get<Money>(row, column)));
+        update_style(*money_cell, apply_table_cell_right_align_style);
+        return money_cell;
+      } else if(column_id == TimeAndSalesTableModel::Column::SIZE) {
+        auto quantity_cell = make_label(
+          to_text(table->get<Quantity>(row, column)).remove(QChar(',')));
+        update_style(*quantity_cell, apply_table_cell_right_align_style);
+        return quantity_cell;
+      } else if(column_id == TimeAndSalesTableModel::Column::MARKET) {
+        return make_label(
+          QString::fromStdString(table->get<std::string>(row, column)));
+      } else if(column_id == TimeAndSalesTableModel::Column::CONDITION) {
+        return make_label(
+          to_text(table->get<TimeAndSale::Condition>(row, column)));
+      }
+      return make_label("");
+    }();
+    update_style(*cell, apply_table_cell_style);
+    return cell;
+  }
+}
+
+TableView* Spire::make_time_and_sales_table_view(
+    std::shared_ptr<TimeAndSalesTableModel> table, QWidget* parent) {
+  auto table_view = TableViewBuilder(table).
+    set_header(make_header_model()).
+    set_view_builder(table_view_builder).make();
+  update_style(*table_view, apply_table_view_style);
+  auto& header = table_view->get_header();
+  auto header_scroll_box = new ScrollBox(&header);
+  header_scroll_box->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Fixed);
+  header_scroll_box->setFocusPolicy(Qt::NoFocus);
+  header_scroll_box->set_horizontal(ScrollBox::DisplayPolicy::NEVER);
+  header_scroll_box->set_vertical(ScrollBox::DisplayPolicy::NEVER);
+  update_style(*header_scroll_box, apply_table_header_style);
+  auto& header_box =
+    *static_cast<Box*>(table_view->layout()->itemAt(0)->widget());
+  auto old_header_box =
+    table_view->layout()->replaceWidget(&header_box, header_scroll_box);
+  delete old_header_box->widget();
+  delete old_header_box;
+  auto properties = make_header_item_properties();
+  for(auto i = 0; i < std::ssize(properties); ++i) {
+    header.get_widths()->set(i, properties[i].m_width);
+    auto item = table_view->get_header().get_item(i);
+    item->setVisible(properties[i].m_is_visible);
+    auto item_layout = item->layout();
+    item_layout->setContentsMargins({scale_width(4), scale_height(5), 0, 0});
+    auto contents_layout =
+      item_layout->itemAt(0)->layout()->itemAt(0)->widget()->layout();
+    contents_layout->setContentsMargins({});
+    if(properties[i].m_alignment == Qt::AlignRight) {
+      static_cast<QSpacerItem*>(contents_layout->itemAt(1))->changeSize(0, 0);
```

Comment:
We need to meet to over this, code like this is not acceptable and has caused a lot of problems. Things need to be coded against APIs, not against implementation details that break encapsulation.

I will approve this PR ignoring this change but we need to meet to discuss this because I've seen this kind of code in many parts of the codebase and it really has resulted in a lot of crashes and maintenance issues.

Reply:
The current `TableHeaderItem` layout doesn't support the right alignment of the label, while the spec update for the `TableHeaderItem` has a different layout that may address the right-alignment issue. I don't want to change it too much since the `TableHeaderItem` should be reimplemented.